File: python/google/protobuf/internal/canonical_api_test.py
#!/usr/bin/env python
# Protocol Buffers - Google's data interchange format
# Copyright 2008 Google Inc.  All rights reserved.
#
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file or at
# https://developers.google.com/open-source/licenses/bsd

"""Tests for the canonical API aliases in Message."""

import unittest

from google.protobuf import unittest_pb2


class CanonicalApiTest(unittest.TestCase):
  """Tests for canonical API method aliases."""

  def test_serialize_equivalence(self):
    """Test that serialize() produces the same output as SerializeToString()."""
    message = unittest_pb2.TestAllTypes()
    message.optional_int32 = 123
    message.optional_string = 'test'

    canonical = message.serialize()
    original = message.SerializeToString()

    self.assertEqual(canonical, original)
    self.assertTrue(len(canonical) > 0)

  def test_parse_equivalence(self):
    """Test that parse() works the same as ParseFromString()."""
    original = unittest_pb2.TestAllTypes()
    original.optional_int32 = 456
    original.optional_string = 'hello'

    data = original.serialize()

    # Test canonical parse()
    parsed_canonical = unittest_pb2.TestAllTypes()
    parsed_canonical.parse(data)
    self.assertEqual(parsed_canonical.optional_int32, 456)
    self.assertEqual(parsed_canonical.optional_string, 'hello')

    # Test original ParseFromString()
    parsed_original = unittest_pb2.TestAllTypes()
    parsed_original.ParseFromString(data)
    self.assertEqual(parsed_original.optional_int32, 456)
    self.assertEqual(parsed_original.optional_string, 'hello')

  def test_clear_equivalence(self):
    """Test that clear() works the same as Clear()."""
    message = unittest_pb2.TestAllTypes()
    message.optional_int32 = 789
    message.optional_string = 'world'

    self.assertTrue(message.HasField('optional_string'))

    message.clear()

    # After clear, default values apply
    self.assertEqual(message.optional_int32, 0)
    self.assertEqual(message.optional_string, '')

  def test_is_initialized_equivalence(self):
    """Test that is_initialized() works the same as IsInitialized()."""
    message = unittest_pb2.TestAllTypes()

    self.assertEqual(message.is_initialized(), message.IsInitialized())
    self.assertTrue(message.is_initialized())

    message.optional_int32 = 100
    self.assertEqual(message.is_initialized(), message.IsInitialized())
    self.assertTrue(message.is_initialized())

  def test_get_serialized_size_equivalence(self):
    """Test that get_serialized_size() works the same as ByteSize()."""
    message = unittest_pb2.TestAllTypes()
    message.optional_int32 = 42
    message.optional_string = 'size test'

    canonical_size = message.get_serialized_size()
    original_size = message.ByteSize()

    self.assertEqual(canonical_size, original_size)
    self.assertTrue(canonical_size > 0)

  def test_merge_from_equivalence(self):
    """Test that merge_from() works the same as MergeFrom()."""
    source = unittest_pb2.TestAllTypes()
    source.optional_int32 = 111

    # Test canonical merge_from()
    target_canonical = unittest_pb2.TestAllTypes()
    target_canonical.optional_string = 'existing'
    target_canonical.merge_from(source)
    self.assertEqual(target_canonical.optional_int32, 111)
    self.assertEqual(target_canonical.optional_string, 'existing')

    # Test original MergeFrom()
    target_original = unittest_pb2.TestAllTypes()
    target_original.optional_string = 'existing'
    target_original.MergeFrom(source)
    self.assertEqual(target_original.optional_int32, 111)
    self.assertEqual(target_original.optional_string, 'existing')

  def test_clone(self):
    """Test that clone() creates a proper copy."""
    original = unittest_pb2.TestAllTypes()
    original.optional_int32 = 12345
    original.optional_string = 'clone test'
    original.repeated_int32.append(1)
    original.repeated_int32.append(2)
    original.repeated_int32.append(3)

    cloned = original.clone()

    # Verify contents
    self.assertEqual(cloned.optional_int32, 12345)
    self.assertEqual(cloned.optional_string, 'clone test')
    self.assertEqual(list(cloned.repeated_int32), [1, 2, 3])

    # Verify it's a separate copy
    cloned.optional_int32 = 99999
    self.assertEqual(original.optional_int32, 12345)

  def test_round_trip(self):
    """Test round-trip: serialize -> parse."""
    original = unittest_pb2.TestAllTypes()
    original.optional_int32 = 12345
    original.optional_int64 = 67890
    original.optional_float = 3.14
    original.optional_string = 'round trip test'
    original.repeated_int32.append(1)
    original.repeated_int32.append(2)
    original.repeated_int32.append(3)

    data = original.serialize()

    restored = unittest_pb2.TestAllTypes()
    restored.parse(data)

    self.assertEqual(restored.optional_int32, 12345)
    self.assertEqual(restored.optional_int64, 67890)
    self.assertAlmostEqual(restored.optional_float, 3.14, places=5)
    self.assertEqual(restored.optional_string, 'round trip test')
    self.assertEqual(list(restored.repeated_int32), [1, 2, 3])

  def test_empty_message(self):
    """Test empty message."""
    empty = unittest_pb2.TestAllTypes()

    data = empty.serialize()
    self.assertEqual(len(data), 0)

    parsed = unittest_pb2.TestAllTypes()
    parsed.parse(data)
    self.assertTrue(parsed.is_initialized())
    self.assertEqual(parsed.get_serialized_size(), 0)

  def test_deterministic_serialization(self):
    """Test deterministic serialization parameter."""
    message = unittest_pb2.TestAllTypes()
    message.optional_int32 = 42

    # Test with deterministic=True
    data1 = message.serialize(deterministic=True)
    data2 = message.serialize(deterministic=True)
    self.assertEqual(data1, data2)


if __name__ == '__main__':
  unittest.main()

File: src/google/protobuf/canonical_api_test.cc
// Protocol Buffers - Google's data interchange format
// Copyright 2008 Google Inc.  All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file or at
// https://developers.google.com/open-source/licenses/bsd

// Tests for the canonical API aliases in MessageLite.

#include "google/protobuf/unittest.pb.h"

#include <string>

#include <gtest/gtest.h>
#include "google/protobuf/message_lite.h"

namespace google {
namespace protobuf {
namespace {

// Test that serialize() produces the same output as SerializeAsString()
TEST(CanonicalApiTest, SerializeEquivalence) {
  protobuf_unittest::TestAllTypes message;
  message.set_optional_int32(123);
  message.set_optional_string("test");

  std::string canonical = message.serialize();
  std::string original = message.SerializeAsString();

  EXPECT_EQ(canonical, original);
  EXPECT_FALSE(canonical.empty());
}

// Test that parse() works the same as ParseFromString()
TEST(CanonicalApiTest, ParseEquivalence) {
  protobuf_unittest::TestAllTypes original;
  original.set_optional_int32(456);
  original.set_optional_string("hello");

  std::string data = original.serialize();

  // Test canonical parse()
  protobuf_unittest::TestAllTypes parsed_canonical;
  EXPECT_TRUE(parsed_canonical.parse(data));
  EXPECT_EQ(parsed_canonical.optional_int32(), 456);
  EXPECT_EQ(parsed_canonical.optional_string(), "hello");

  // Test original ParseFromString()
  protobuf_unittest::TestAllTypes parsed_original;
  EXPECT_TRUE(parsed_original.ParseFromString(data));
  EXPECT_EQ(parsed_original.optional_int32(), 456);
  EXPECT_EQ(parsed_original.optional_string(), "hello");
}

// Test that clear() works the same as Clear()
TEST(CanonicalApiTest, ClearEquivalence) {
  protobuf_unittest::TestAllTypes message;
  message.set_optional_int32(789);
  message.set_optional_string("world");

  EXPECT_TRUE(message.has_optional_int32());
  EXPECT_TRUE(message.has_optional_string());

  message.clear();

  EXPECT_FALSE(message.has_optional_int32());
  EXPECT_FALSE(message.has_optional_string());
}

// Test that isInitialized() works the same as IsInitialized()
TEST(CanonicalApiTest, IsInitializedEquivalence) {
  protobuf_unittest::TestAllTypes message;

  // TestAllTypes has no required fields in proto3
  EXPECT_EQ(message.isInitialized(), message.IsInitialized());
  EXPECT_TRUE(message.isInitialized());

  message.set_optional_int32(100);
  EXPECT_EQ(message.isInitialized(), message.IsInitialized());
  EXPECT_TRUE(message.isInitialized());
}

// Test that getSerializedSize() works the same as ByteSizeLong()
TEST(CanonicalApiTest, GetSerializedSizeEquivalence) {
  protobuf_unittest::TestAllTypes message;
  message.set_optional_int32(42);
  message.set_optional_string("size test");

  size_t canonical_size = message.getSerializedSize();
  size_t original_size = message.ByteSizeLong();

  EXPECT_EQ(canonical_size, original_size);
  EXPECT_GT(canonical_size, 0u);
}

// Test that mergeFrom() works the same as MergeFromString()
TEST(CanonicalApiTest, MergeFromEquivalence) {
  protobuf_unittest::TestAllTypes source;
  source.set_optional_int32(111);

  std::string data = source.serialize();

  // Test canonical mergeFrom()
  protobuf_unittest::TestAllTypes target_canonical;
  target_canonical.set_optional_string("existing");
  EXPECT_TRUE(target_canonical.mergeFrom(data));
  EXPECT_EQ(target_canonical.optional_int32(), 111);
  EXPECT_EQ(target_canonical.optional_string(), "existing");

  // Test original MergeFromString()
  protobuf_unittest::TestAllTypes target_original;
  target_original.set_optional_string("existing");
  EXPECT_TRUE(target_original.MergeFromString(data));
  EXPECT_EQ(target_original.optional_int32(), 111);
  EXPECT_EQ(target_original.optional_string(), "existing");
}

// Test round-trip: serialize -> parse
TEST(CanonicalApiTest, RoundTrip) {
  protobuf_unittest::TestAllTypes original;
  original.set_optional_int32(12345);
  original.set_optional_int64(67890);
  original.set_optional_float(3.14f);
  original.set_optional_string("round trip test");
  original.add_repeated_int32(1);
  original.add_repeated_int32(2);
  original.add_repeated_int32(3);

  std::string data = original.serialize();

  protobuf_unittest::TestAllTypes restored;
  EXPECT_TRUE(restored.parse(data));

  EXPECT_EQ(restored.optional_int32(), 12345);
  EXPECT_EQ(restored.optional_int64(), 67890);
  EXPECT_FLOAT_EQ(restored.optional_float(), 3.14f);
  EXPECT_EQ(restored.optional_string(), "round trip test");
  EXPECT_EQ(restored.repeated_int32_size(), 3);
  EXPECT_EQ(restored.repeated_int32(0), 1);
  EXPECT_EQ(restored.repeated_int32(1), 2);
  EXPECT_EQ(restored.repeated_int32(2), 3);
}

// Test empty message
TEST(CanonicalApiTest, EmptyMessage) {
  protobuf_unittest::TestAllTypes empty;

  std::string data = empty.serialize();
  EXPECT_TRUE(data.empty());

  protobuf_unittest::TestAllTypes parsed;
  EXPECT_TRUE(parsed.parse(data));
  EXPECT_TRUE(parsed.isInitialized());
  EXPECT_EQ(parsed.getSerializedSize(), 0u);
}

}  // namespace
}  // namespace protobuf
}  // namespace google
